portfolio/src/main/java/com/google/sps/data/BlogHashMap.java

```diff
@@ -0,0 +1,175 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public final class BlogHashMap {
+  private static final String[] internalTags = {"#general", "#wellbeing", "#music", "#games"};
+
+  // Takes and puts a list of BlogMessage type in map.
+  // Each BlogMessage type should contain tag that matches tags in the internal System. 
+  public static void putInMap(List<BlogMessage> messages, Map<String, LinkedList<BlogMessage>> map) {
+    for (BlogMessage message : messages) {
+      boolean isTagSupported = false;
+      for (int i = 0; i < internalTags.length; i++){
+        if (message.getTag().equals(internalTags[i])){
+          isTagSupported = true;
+          if (map.containsKey(message.getTag())){
+            map.get(message.getTag()).addLast(message);
+          } else {
+            // No such mapping with specified tag,
+            // Create first mapping with specified tag.
+            LinkedList<BlogMessage> firstValue = new LinkedList<BlogMessage>();
+            firstValue.addFirst(message); 
+            map.put(message.getTag(), firstValue);
+          }
+        } 
+      }
+      if (!isTagSupported) {
+        System.out.println("Tag not supported. Tag: " + message.getTag());
+        System.out.println("For BlogMessage@ " + message);
+      }
+    }
+
+  }
+
+  public static int getSize(Map<String, LinkedList<BlogMessage>> map) {
+    return map.size();
+  }
+
+  // Takes map, and
+  // Returns a linkedList of all values of BlogMessage type in the map. 
+  public static LinkedList<BlogMessage> getMessages(Map<String, LinkedList<BlogMessage>> map) {
+    System.out.println("Fetching all [LinkedList] Values in map...");
+
+    Set<String> keys = map.keySet();
+    LinkedList<BlogMessage> values = new LinkedList<BlogMessage>();
+    for (String key : keys) {
+      values.addAll(map.get(key));
+    }
+    return values;
+  }
+
+  // Returns a requested amount of all BlogMessages in map
+  public static LinkedList<BlogMessage> getMessages(Map<String, LinkedList<BlogMessage>> map, int limit) {
+    if (limit < 0) {
+      System.out.println("Cannot load this amount.");
+      return new LinkedList<>();
+    }
+
+    LinkedList<BlogMessage> values = getMessages(map);
+    if (limit > values.size()) {
+      limit = values.size();
+    } 
+
+    System.out.println("Fetching requested amount of [LinkedList] Values for all tags...");
+
+    Iterator<BlogMessage> iterateList = values.iterator();
```

Comment:
Line 87-94 looks a bit inefficient because you call getMessages in line 80, copying all the items to values already, and then you are trying to copy the portion within limit twice, and would have a waste in copying the rest in the first call to getMessages(). Try to see if you can improve that.

Reply:
fron what I understand you mean getting all items in the LinkedList and then copying a portion into the limitedLinkedList is double job, and the rest that are not copied will basically be useless?
So instead of first getting the original values and then copying part of it I could just get the requested amount straight from the hashmap. Is that close to addressing your comment?

Reply:
Exactly! This usually matters when the HashMap is big, cause the time complexity is bound to the size of the internal hashmap, not the amount the caller requested, i.e., even if I just need one item (limit = 1), I would need to copy the whole hashmap, and then return the first one:)

Reply:
Done!